lib/web/forms/registration.go

```diff
@@ -0,0 +1,71 @@
+package webforms
+
+import (
+	"errors"
+
+	"go-google-scraper-challenge/constants"
+	"go-google-scraper-challenge/helpers"
+	"go-google-scraper-challenge/lib/models"
+
+	validation "github.com/go-ozzo/ozzo-validation/v4"
+	"github.com/go-ozzo/ozzo-validation/v4/is"
+)
+
+type RegistrationForm struct {
+	Email                string `form:"email"`
+	Password             string `form:"password"`
+	PasswordConfirmation string `form:"password_confirmation"`
+}
+
+func (f RegistrationForm) Validate() (valid bool, err error) {
+	err = validation.ValidateStruct(&f,
+		validation.Field(&f.Email, validation.Required, is.EmailFormat),
+		validation.Field(&f.Password, validation.Required),
+		validation.Field(&f.PasswordConfirmation, validation.Required.When(f.Password != ""), validation.By(f.validatePasswordConfirmation(f.Password))),
+	)
+
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (f RegistrationForm) validatePasswordConfirmation(password string) validation.RuleFunc {
+	return func(value interface{}) error {
+		if value.(string) != password && len(password) > 0 {
+			return errors.New("does not match the password")
+		}
+
+		return nil
+	}
+}
+
+func (f RegistrationForm) Save() (*models.User, error) {
+	existingUser, err := models.GetUserByEmail(f.Email)
+	if err == nil && existingUser != nil {
+		return nil, errors.New(constants.UserAlreadyExist)
+	}
+
+	hashedPassword, err := helpers.HashPassword(f.Password)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
@carryall I see that gorm has a [hook](https://gorm.io/docs/hooks.html#Hooks).
I think you could use that for a chance to try on that feature.
It's OK to me too if use this 🖖

Reply:
Great idea 👍🏼 I updated in [a77ef4c](https://github.com/carryall/go-google-scraper-challenge/pull/125/commits/a77ef4cf15a54ca2e8f845ac48826ca2bfcb9f05)